### PR TITLE
Machine/Power Code Refactor + Spatial Grid System + Performance Updates

### PR DESCRIPTION
**Pull request by DoctorSquishy**
# About The Pull Request
My biggest PR yet! **Wait til the big cooking port is in, so I can update their power code with this.**
Ports optimizations of machine code from TG. For more details read: https://github.com/tgstation/tgstation/pull/60317
- Added SI unit formatting proc and applied it to some examines
- Refactors machine_stat and is_processing() to process on demand 
- Start and stop processing depending on if they are operational, instead of always processing regardless.
- Large refactor of machine/power code to cut down on processing time and wasted lists
- Direction Control of Vents using the AirAlarm
- Refactors occupant to use setter
- Introduces Spatial Grid System
- Refactors Lights
- New Alarm Handler Datum

## Why It's Good For The Game
QoL optimizations of machine code and power processing on demand
Better callbacks for code

## Testing Procedure
Running through machines and using them as the station runs.

## PR References
- https://github.com/tgstation/tgstation/pull/60317 
(The **BIG** KAHUNA) 
- https://github.com/tgstation/tgstation/pull/52235 
(Auto SI Units without TGUI, good for examines)
- https://github.com/tgstation/tgstation/pull/54600 
(Auto SI Units without TGUI)
- https://github.com/tgstation/tgstation/pull/53150 
(Finish from what I started with machine_stat)
- https://github.com/tgstation/tgstation/pull/54373 
(Part of machines)
- https://github.com/tgstation/tgstation/pull/51536 
(While updating AirAlarms why not)
- https://github.com/tgstation/tgstation/pull/67497
(To help prevent Overlay Overflows, and assist with future scrubber optimizations that process when needed)
- https://github.com/tgstation/tgstation/pull/60060
(New Alarm Handler)
- https://github.com/tgstation/tgstation/pull/61422
(Spatial Grid System)

- https://github.com/tgstation/tgstation/pull/52970
- https://github.com/tgstation/tgstation/pull/53943

## Changelog

:cl: DoctorSquishy, Donkie, MadMedicineMan, Rohesie, Kylerace, nicbn, WarlockD
tweak: allows the air alarm to let vents suck are as well as pressurize.
code: Machine code should run smoother now which means less server lag!
/:cl:
add: New Light Switches
imageadd: New Light Switch Icons
refactor: Light Refactor
add: Air Alarms have an option for Pressure Vents to siphon
fix: Fixed energy_ball balls

### CODE REVIEW

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm [hunk 1]
```diff
@@ -173,13 +182,16 @@
...
 
-/obj/machinery/atmospherics/components/unary/cryo_cell/process(delta_time)
-	if(!on)
-		return
-
-	if(!is_operational())
+/obj/machinery/atmospherics/components/unary/cryo_cell/on_set_is_operational(old_value)
+	if(old_value) //Turned off
 		on = FALSE
+		end_processing()
```
Comment (nednaZ) [inline code/modules/atmospherics/machinery/components/unary_devices/cryo.dm:188]: I had actually planned for this myself, disabling processing when not in use

**Review by nednaZ — CHANGES_REQUESTED** [review]
Nothing too bad

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm (hunk 1 above)
Comment (DoctorSquishy) [reply code/modules/atmospherics/machinery/components/unary_devices/cryo.dm:188]: If there is other machines you think I should look at for this kind of change, I can see about adding it.

File: code/controllers/subsystem/overlays.dm [hunk 2]
```diff
@@ -38,24 +38,48 @@ SUBSYSTEM_DEF(overlays)
...
-			STAT_LOG_ENTRY(stats, A.type)
+		if(!atom_to_compile)
+			continue
+		STAT_START_STOPWATCH
+		COMPILE_OVERLAYS(atom_to_compile)
+		UNSETEMPTY(atom_to_compile.add_overlays)
+		UNSETEMPTY(atom_to_compile.remove_overlays)
+		STAT_STOP_STOPWATCH
+		STAT_LOG_ENTRY(stats, atom_to_compile.type)
+		if(length(atom_to_compile.overlays) >= MAX_ATOM_OVERLAYS)
```
[the remark this answers is not in the document]
Comment (DoctorSquishy) [reply code/controllers/subsystem/overlays.dm:51]: https://github.com/tgstation/tgstation/pull/67497 This PR moved it there, so will be it okay to move it back above it?

File: code/game/machinery/_machinery.dm [hunk 3]
```diff
@@ -104,23 +106,40 @@ Class Procs:
 		//0 = dont run the auto
```
[the remark this answers is not in the document]
Comment (DoctorSquishy) [reply code/game/machinery/_machinery.dm]: Done

File: code/_compile_options.dm [hunk 4]
```diff
@@ -108,6 +108,10 @@
 #define CBT
 #endif
 
+// A reasonable number of maximum overlays an object needs
+// If you think you need more, rethink it
+#define MAX_ATOM_OVERLAYS 100
```
[the remark this answers is not in the document]
Comment (DoctorSquishy) [reply code/_compile_options.dm]: I went ahead and moved it to the DEFINES folder

**Comment by Zonespace27** [comment]
edit the power sink requirement to blow up, atm on TG it's just a syndicate bomb+ that explodes after two minutes at most

**Comment by nednaZ** [comment]
Air alarms no longer go into their alert mode sprite when the air is unsafe

**Comment by nednaZ** [comment]
Looks to be working as intended, at least until it hits more test merging!

**Comment by nednaZ** [comment]
Machines that are deconstructed are not dropping boards.

**Comment by DoctorSquishy** [comment]
> Machines that are deconstructed are not dropping boards.

Fixed, TG had some slight differences on handling circuits in machines so I updated it

**Review by nednaZ — APPROVED** [review]
After testmerging and some bugfixes here and there, we haven't found any further major issues.
The machines SS runs faster than it was, which is good in my eyes.

I'd suggest offering this to Bee main as well, they may like the performance boost.

**Pull request merged**